### PR TITLE
refactor routers/router.go

### PR DESCRIPTION
**Pull request by nickhsine**
@babygoat 

[first commit f8bd86a](https://github.com/twreporter/go-api/commit/f8bd86aba5535a5d555bad95a9a1dd630108b88e) has been already reviewed in `feature-donations` branch. You can just review the second commit.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -30,45 +28,39 @@ func main() {
...
 		panic(err)
 	}
 
-	defer utils.Check(cf.Close)
+	// mailSender := utils.NewSMTPEmailSender()                          // use office365 to send mails
+	mailSender := utils.NewAmazonEmailSender() // use Amazon SES to send mails
+
+	cf = controllers.NewControllerFactory(db, session, mailSender)
+
+	defer db.Close()
```
Comment (babygoat) [inline main.go]: if mongodb initializes failed, then db.close will not be executed.

**Comment by babygoat** [comment]
Others, LGTM

File: routers/router.go [hunk 2]
```diff
@@ -101,5 +105,23 @@ func SetupRouter(cf *controllers.ControllerFactory) *gin.Engine {
...
 
+	// =============================
+	// v2 oauth endpoints
+	// =============================
+	v2Group := engine.Group("/v2")
+	v2AuthGroup := v2Group.Group("/auth")
+
+	session := cf.GetMgoSession()
+	c := session.DB("go-api").C("sessions")
+	store := mongo.NewStore(c, 3600, true, []byte("secret"))
```
Comment (nickhsine) [inline routers/router.go]: `secret` needs to move into config

File: models/model_user.go [hunk 3]
```diff
@@ -2,11 +2,49 @@ package models
...
+		return json.Marshal(nil)
+	}
+}
+
+func (v *NullString) UnmarshalJSON(data []byte) error {
+	var s *string
+	if err := json.Unmarshal(data, &s); err != nil {
+		return err
+	}
+	log.Infof("s: %#v", *s)
```
Comment (nickhsine) [inline models/model_user.go]: remove `log.Infof`

File: routers/router.go (hunk 2 above)
Comment (nickhsine) [reply routers/router.go]: increase `maxAge` value from `3600`(2 hours) to `604800` (1 week)

File: controllers/oauth.go [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+func beginAuth(c *gin.Context, conf *oauth2.Config) {
+	var state string
+	var err error
+
+	destination := c.Request.URL.Query().Get("destination")
+	if destination == "" {
+		destination = defaultDestination
+	}
+
+	log.Infof("destination beginAuth %s", destination)
```
Comment (nickhsine) [inline controllers/oauth.go]: remove `log.info`

File: controllers/oauth.go [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+		state = "twreporter-oauth-state"
+	}
+
+	session := sessions.Default(c)
+	session.Set("state", state)
+	session.Set("destination", destination)
+	session.Save()
+
+	url := conf.AuthCodeURL(state)
+	log.Infof("url %#v", url)
```
Comment (nickhsine) [inline controllers/oauth.go]: remove `log.Info`

Comment (babygoat) [inline controllers/oauth.go]: Remove?

File: controllers/oauth.go [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+func getOauthUserInfo(c *gin.Context, conf *oauth2.Config, userInfoEndpoint string, oauthUser interface{}) error {
+	session := sessions.Default(c)
+	retrievedState := session.Get("state")
+	state := c.Request.URL.Query().Get("state")
+	if state != retrievedState {
+		log.Warnf("expect state is %s, but actual state is %s", retrievedState, state)
+		return models.NewAppError("getOauthUserInfo", "oauth fails", "Invalid oauth state", 500)
+	}
+
+	code := c.Request.URL.Query().Get("code")
```
Comment (babygoat) [inline controllers/oauth.go]: c.Query("code")?

File: routers/router.go (hunk 2 above)
Comment (babygoat) [reply routers/router.go]: I suggest use constant `const maxAge = 3600` for magic number

Comment (babygoat) [reply routers/router.go]: I suggest using constant literal rather than magic number. (i.e., `const maxAge=3600`)

File: controllers/oauth.go [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+
+// beginAuth uses sessions to store users'
+// 1. state
+// 2. destination(go to page)
+// and redirect users to oauth server.
+func beginAuth(c *gin.Context, conf *oauth2.Config) {
+	var state string
+	var err error
+
+	destination := c.Request.URL.Query().Get("destination")
```
Comment (babygoat) [inline controllers/oauth.go]: c.Query("destination")

File: controllers/oauth.go [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+}
+
+// getOauthUserInfo does the following three things
+// 1. validate state
+// 2. exchange code to token
+// 3. get user info from oauth server by token
+func getOauthUserInfo(c *gin.Context, conf *oauth2.Config, userInfoEndpoint string, oauthUser interface{}) error {
+	session := sessions.Default(c)
+	retrievedState := session.Get("state")
+	state := c.Request.URL.Query().Get("state")
```
Comment (babygoat) [inline controllers/oauth.go]: c.Query("state")

File: main.go (hunk 1 above)
Comment (nickhsine) [reply main.go]: will move `defer` function before `panic`

**Pull request merged**